[MM-22846]: post forwarding

michelengelen opened the pull request:
#### Summary
this adds post forwarding according to [this design](https://www.figma.com/file/LQr3Y0Fmbe1vxuL99Pt2A9/MM-22846-Forward-Messages?node-id=5752%3A63706)

#### Ticket Link
[MM-22846](https://mattermost.atlassian.net/browse/MM-22846)

#### Related Pull Requests
- Has server changes https://github.com/mattermost/mattermost-server/pull/20166

#### Screenshots
tbd

#### Release Note
```release-note
adds post forwarding feature
```

esethna commented:
Thanks @michelengelen!

A few things I noticed:

1. We should allow CTRL + ENTER to send the forwarded message when in the forward modal. Similarly, when not focused in the comment field, allow ENTER to send the forward message if a destination channel has been selected
2. Somethings strange in the autocomplete when searching
![image](https://user-images.githubusercontent.com/14333569/174651890-f2d226f9-e739-46dd-9ca8-48d1e7664048.png)
3. Button styles for the forward and cancel buttons don't look quite right, I'll defer to @matthewbirtch 
4. Hover effect on the "x" doesn't look quite right missing rounded corners, I'll defer to @matthewbirtch 
5. Styling for long messages overruns modal
![image](https://user-images.githubusercontent.com/14333569/174652419-ce6b2fc2-842a-495f-bbb5-537d5c414523.png)
6. Forward button is disabled in DMs and GMs
![image](https://user-images.githubusercontent.com/14333569/174652592-9a890103-30d4-467d-bb91-009f6aee4308.png)

esethna commented:
@hmhealey @BenCookie95 @matthewbirtch would you guys mind expediting reviews if possible? Thanks!

matthewbirtch reviewed (changes requested):
Nice work on this @michelengelen. A few UI things from me in addition to what was found by @esethna:

1. Can we update the modal header to use the same header as the `create channel` modal? The `x` icon button style/icon/hover state looks to be outdated and positioned awkwardly
2. Can we update the footer buttons to use the up-to-date button components that are used in the `create channel` modal?
3. I'm seeing some inconsistencies with the select channel input and the `add comment` input. Different padding and different opacities on the text color
<img width="649" alt="Screen Shot 2022-06-20 at 1 36 34 PM" src="https://user-images.githubusercontent.com/2040554/174655446-93880a25-f018-4206-8d9b-8ef70cf56518.png">

5. The focus state of the `comment` input isn't right either.

4. We should remove the hover state on the message preview since it's not clickable in this case
5. I didn't get an error message if I try to forward without selecting a channel
6. The `Forward` button should be disabled until a channel is selected. Visually it looks enabled right away. Fixing this may resolve the issue above since it will prevent errors
8. With long posts, if I click the 'Show more' it expands beyond the height of the modal and overlaps the buttons in the footer. I'm actually thinking that we might be able to just remove the 'show more' button in this context to prevent this. It feels unnecessary to have the show more when forwarding. Alternatively, we could make sure it just scrolls properly within the main content area rather than overlapping the footer.
<img width="660" alt="image" src="https://user-images.githubusercontent.com/2040554/174654612-1a979911-d0ad-4f1d-8a2b-6b50138be460.png">

9. This hover actions menu is getting pretty full now. @esethna I'm wondering if we should move this to the ••• menu instead? There are 9 items here now
<img width="295" alt="image" src="https://user-images.githubusercontent.com/2040554/174655248-958eb2bd-e0b3-4abd-a9e0-c0fd1c6283ea.png">

7. Are we going to make this function available in the RHS as well? The space is much more limited there, so it would likely need to go in the ••• menu in this context anyway

esethna commented:
Re 9 + 10: @matthewbirtch, I'm really worried about discoverability if we move it to the ... menu. I'd propose we leave it as is and also add it to the RHS (there seems to be enough room even on small screens). We can always adjust if we get feedback on it being too long or hard to consume

![image](https://user-images.githubusercontent.com/14333569/174837781-ff46f946-a47a-40d6-96cd-c2ef61f4c13b.png)

11. @matthewbirtch I also realized we never designed a "tutorial tip" for this. Should we add something similar to the actions menu tip?

matthewbirtch commented:
> 

@esethna what do you think about removing the 'reply' icon from this hover menu since it's in the ••• menu already and we have a few other methods to reply (click to open thread, clicking 'reply' in the thread footer)

esethna commented:
@matthewbirtch I would rather keep reply (even though it may be repetitive it's the most critical functionality to surface). If we move the forward to the ... menu then let's make sure we design a tutorial tip to help discoverability

matthewbirtch commented:
> @matthewbirtch I would rather keep reply (even though it may be repetitive it's the most critical functionality to surface - and some users might not have the click to reply on). If we move the forward to the ... menu then let's make sure we design a tutorial tip to help discoverability

Ya, I don't disagree with that. I'll make a ticket to design something to bring attention to the feature in the ••• menu

matthewbirtch reviewed (changes requested):
@michelengelen circling back on where this features should live. After speaking with a number of team members I think we should move it out of the hover menu since it's getting too crowded and to move it in to the ••• menu. 

I've updated the Figma file to include this change: https://www.figma.com/file/LQr3Y0Fmbe1vxuL99Pt2A9/MM-22846-Forward-Messages?node-id=6224%3A105165

This also includes how we can introduce this feature using a pulsing dot on the ••• button and a `NEW` badge on the menu item itself

michelengelen commented:
@matthewbirtch and @esethna ... with my latest changes all of your suggestions should now be addressed. Would you mind reviewing again?

The only thing I did not add yet is the pulsating dot and the new badge. Will add this later today or tomorrow.

BTW: using the `GenericModal` did add some complexities I did not expect, that's why it took longer than expected to adjust! Sorry! 🙇🏼

matthewbirtch reviewed (changes requested):
@michelengelen looks like a lot of the issues are resolved. Thanks! A few new ones I discovered this round:
1. When a message preview has an image in it, the modal seems to be calculating its position incorrectly and displaying at the bottom rather than centered. 
<img width="2032" alt="image" src="https://user-images.githubusercontent.com/2040554/175363951-495cf8d0-6f2b-42ac-8065-b81ad77197cd.png">
2. When the select menu is open it is showing in behind image attachments in the message preview
<img width="680" alt="image" src="https://user-images.githubusercontent.com/2040554/175364170-fc5415e5-3284-4d64-b087-8a16a859ee07.png">

3. Long posts have the 'Show more' visible, but clicking it doesn't do anything. Are we still going to remove this in this context?

esethna commented:
Looks like we need a keyboard shortcut for the ... menu, maybe "S" for share? 0/5 @matthewbirtch?

![image](https://user-images.githubusercontent.com/14333569/175644045-dd7c20d7-e8c6-4626-a8ce-343f52dfb918.png)

esethna reviewed (approved):
Looks good otherwise, thanks Michel!

matthewbirtch commented:
> Looks like we need a keyboard shortcut for the ... menu, maybe "S" for share? 0/5 @matthewbirtch?

Good catch, I don't have a great suggestion for the shortcut since `F` is already taken. `S` works as good as anything else I guess.

michelengelen commented on packages/mattermost-redux/src/selectors/entities/preferences.ts:
```diff
@@ -199,6 +199,11 @@ export function isCollapsedThreadsEnabled(state: GlobalState): boolean {
 
+// TODO: remove the `true` alternate return when the feature is tested and passed on the PR
+export function getIsPostForwardingEnabled(state: GlobalState): boolean {
+    return getFeatureFlagValue(state, 'PostForwarding') === 'true' || true;
```
same here ... i do have permission, but i am super lost on their UI :D

Willyfrog replied:
feel free to ping me, not super savvy on that but I can at least explain most of its ui.

michelengelen commented on components/suggestion/switch_channel_provider.jsx:
```diff
@@ -424,7 +424,7 @@ export default class SwitchChannelProvider extends Provider {
     }
 
-    async fetchUsersAndChannels(channelPrefix, resultsCallback) {
+    async fetchUsersAndChannels(channelPrefix, resultsCallback, forceDispatch = false) {
```
its not needed anymore ... will remove

michelengelen reviewed (commented):
Thanks @hmhealey for the review ... I have left some comments on your comments and fixed what was possible!

michelengelen commented:
> > Looks like we need a keyboard shortcut for the ... menu, maybe "S" for share? 0/5 @matthewbirtch?
> 
> Good catch @esethna, I don't have a great suggestion for the shortcut since `F` is already taken. `S` works as good as anything else I guess.

@matthewbirtch shortcut `S` is already taken (used for saving/unsaving). We could either find a new one, or start using combinations, e.g. `Shift + F` or something of the sort.

cc @esethna

matthewbirtch commented:
> @matthewbirtch shortcut `S` is already taken (used for saving/unsaving). We could either find a new one, or start using combinations, e.g. `Shift + F` or something of the sort.

Not against the idea of going with `Shift` `F`, but I was hoping to keep it to a single letter for consistency in this menu. Any other single-letter shortcut will be fairly arbitrary, but we could use the second letter of the word and use `O`

michelengelen commented:
> > @matthewbirtch shortcut `S` is already taken (used for saving/unsaving). We could either find a new one, or start using combinations, e.g. `Shift + F` or something of the sort.
> 
> Not against the idea of going with `Shift` `F`, but I was hoping to keep it to a single letter for consistency in this menu. Any other single-letter shortcut will be fairly arbitrary, but we could use the second letter of the word and use `O`

`O` might also imply using first character of the word `Onward` for it! ... so that might still work in that way ... kinda! :D

esethna commented:
@matthewbirtch separate from this effort, should we consider adding save to the ... menu even though it's also in the hover menu? I had no idea "S" worked since the shortcut isn't shown anywhere, and it's a pretty commonly used feature I would imagine

matthewbirtch commented:
> > > @matthewbirtch shortcut `S` is already taken (used for saving/unsaving). We could either find a new one, or start using combinations, e.g. `Shift + F` or something of the sort.
> > 
> > 
> > Not against the idea of going with `Shift` `F`, but I was hoping to keep it to a single letter for consistency in this menu. Any other single-letter shortcut will be fairly arbitrary, but we could use the second letter of the word and use `O`
> 
> `O` might also imply using first character of the word `Onward` for it! ... so that might still work in that way ... kinda! :D

The only problem is that it might get confused with `0`

matthewbirtch commented:
> separate from this effort, should we consider adding save to the ... menu even though it's also in the hover menu? I had no idea "S" worked since the shortcut isn't shown anywhere, and it's a pretty commonly used feature I would imagine

@esethna, I wouldn't be opposed to that. Would you mind creating a ticket?

michelengelen commented:
> > > > @matthewbirtch shortcut `S` is already taken (used for saving/unsaving). We could either find a new one, or start using combinations, e.g. `Shift + F` or something of the sort.
> > > 
> > > 
> > > Not against the idea of going with `Shift` `F`, but I was hoping to keep it to a single letter for consistency in this menu. Any other single-letter shortcut will be fairly arbitrary, but we could use the second letter of the word and use `O`
> > 
> > 
> > `O` might also imply using first character of the word `Onward` for it! ... so that might still work in that way ... kinda! :D
> 
> The only problem is that it might get confused with `0`

To check this I will add 2 options ... one with `Shift + F`, the other one with `X` for cross-posting! (really fishy explanation, aye? :D)

matthewbirtch commented:
> To check this I will add 2 options ... one with `Shift + F`, the other one with `X` for cross-posting! (really fishy explanation, aye? :D)

Thanks for providing the options @michelengelen. I think my choice would be `Shift+F`. X feels a bit random.

esethna commented:
Adding @cwarnermm for review of the menu text

BenCookie95 commented on components/forward_post_modal/forward_post_modal.tsx:
```diff
@@ -0,0 +1,384 @@
+                        value={selectedChannel}
+                    />
+                )}
+                <Textbox
```
This component is pretty big and it's doing a lot, can we further abstract it by moving some stuff into smaller components? Moving Textbox and it's functions might help

matthewbirtch commented:
Still seeing these issues @michelengelen: 

- modal opens too low when image attachment preview is in shared message
<img width="2032" alt="image" src="https://user-images.githubusercontent.com/2040554/176276978-134d502f-76ac-45c4-b90b-3cedaac47830.png">

Menu sits under image attachment preview
<img width="669" alt="image" src="https://user-images.githubusercontent.com/2040554/176277068-622a194b-bd3a-4db6-b48b-3889a81faea4.png">

cwarnermm commented:
Can the **Forward Message** option available from **...** be sentence case: "Forward message". It'll match the follow action. Over time, docs plans to sentence case the other options as well for consistency. Also provided casing feedback in the Figma file directly @matthewbirtch

matthewbirtch commented:
@michelengelen I'm still seeing the modal sitting too low when an image is in the preview:
<img width="1599" alt="image" src="https://user-images.githubusercontent.com/2040554/176479213-6c17f210-8c38-4203-b19e-2666e3eebe3a.png">

cwarnermm commented on i18n/en.json:
```diff
@@ -3217,6 +3217,15 @@
+  "forward_post_modal.notification.private_channel": "This message is from a private channel and can only be shared with <strong>{channel}</strong>",
+  "forward_post_modal.preview.footer_message": "Originally posted in ~{channel}",
+  "forward_post_modal.preview.title": "Message preview",
+  "forward_post_modal.title": "Forward message",
```
This may be a case where we're changing from "Forward message" to "Forward"

cwarnermm reviewed (commented):
Feedback provided inline for a string that may need to change from "Forward message" to "Forward" CC @matthewbirtch

BenCookie95 commented on components/forward_post_modal/forward_post_modal.scss:
```diff
@@ -0,0 +1,137 @@
+    }
+}
+
+.console__body .modal .GenericModal .modal-body .form-control,
```
It does effect the styles of other modals, you can see an example on the `Create new channel` modal

michelengelen replied:
interesting ... it should not do that! ... will change it to be more specific about this modal

hmhealey commented on components/forward_post_modal/forward_post_channel_select.tsx:
```diff
@@ -0,0 +1,397 @@
+    let description = `~${details.name}`;
+
+    let tag = null;
+    if (details.type === Constants.DM_CHANNEL) {
```
Interesting, is that a Channels team thing? I haven't heard of doing anything like that, but it definitely sounds like a good idea.

I'd be fine leaving that for later though

michelengelen replied:
we have it for the first time now to see if it reduces tech debt considerably.

hmhealey commented on components/forward_post_modal/forward_post_channel_select.tsx:
```diff
@@ -0,0 +1,397 @@
+            ];
+        };
+
+        provider.fetchAndFormatRecentlyViewedChannels(handleDefaultResults);
```
The logic between the two works somewhat the same, but not exactly the same, and that's when you end up with a lot of branching logic that makes the code harder to follow and reason about. It also means that a change to one feature can accidentally affect another one, like if we make the channel switcher sorting configurable, we probably don't want that to apply here as well.

As for the abusing logic part, the suggestion components are more complicated because they need to support things like matching only part of the entered text (as opposed to all the text here and in the channel switcher), and they also have logic for trying to replace partial autocomplete terms when an entry is selected. Neither of those are needed in the channel switcher, and there's been quite a few bugs in the channel switcher caused by those specifically. Also, they share CSS for the autocomplete list which causes even more issues.

And for this case, using a Provider without the rest of the suggestion components feels a bit like a less extreme version of using something like `React.Component` without the rest of `React` somehow. Yeah, you might be able to reuse some logic from it, but you're using it in a way it wasn't intended to work, and it might break down the line because of that

michelengelen replied:
ok, fair enough ... you convinced me to write my own logic and replace this with that. 👍🏼

hmhealey replied:
I guess this didn't end up happening then? At this point, I'd be okay letting this go in even though I'm still worried about this causing more problems down the line

hmhealey commented on components/forward_post_modal/forward_post_channel_select.tsx:
```diff
@@ -0,0 +1,397 @@
+            };
+
+            provider.handlePretextChanged(inputValue, handleResults);
+            resolve(options);
```
It's a bit of a legacy thing since I wrote the original providers before we used Redux, and I also didn't intend for them to mix results from the client and server. That's also why the providers aren't really connected to Redux properly and why they don't store their results in their either

michelengelen replied:
just another reason to write my own logic around that... might take a while though, so stay tuned for changes! :D

hmhealey commented on components/forward_post_modal/forward_post_modal.tsx:
```diff
@@ -0,0 +1,434 @@
+    const selectedChannelId = selectedChannel?.details?.id || '';
+
+    const canPostInSelectedChannel = useSelector((state: GlobalState) => Boolean(selectedChannelId) && haveIChannelPermission(state, selectedChannel?.details?.team_id || '', selectedChannelId, Permissions.CREATE_POST));
+    const useChannelMentions = useSelector((state: GlobalState) => haveIChannelPermission(state, selectedChannel?.details?.team_id || '', selectedChannelId, Permissions.USE_CHANNEL_MENTIONS));
```
I'd be fine moving either moving them all to the index or here. I still prefer the `connect` method myself for keeping the redux code out of the component, but that's a personal preference thing.

I do still think the design of showing the user channels that they can't forward the post to is a bit weird though from a user perspective

hmhealey commented on components/forward_post_modal/forward_post_modal.tsx:
```diff
@@ -0,0 +1,434 @@
+        if (currentChannel.type === General.PRIVATE_CHANNEL) {
+            const channel = `~${currentChannel.display_name}`;
+            notificationText = (
+                <FormattedMarkdownMessage
```
Hmm, that's a good question. Even without the strong tags, that actually wouldn't work since the `and` wouldn't get translated. I think we might be able to do something like using two different translation strings, one for a single user and one for multiple sort of like we do it for the [`CombinedSystemMessge`'s "userA and userB have joined the channel"](https://github.com/mattermost/mattermost-webapp/blob/master/components/post_view/combined_system_message/combined_system_message.tsx#L24). Then, instead of passing in a string for `participants`, then we could pass in a React object that looks like
```js
<FormattedMessage
    defaultMessage='Forwarded to {participants} and {lastParticipant}.'
    values={{
        participants: (
            <>
                <strong>{'user1'}</strong>
                {', '}
                <strong>{'user2'}</strong>
            <>
        ),
        lastParticipant: <strong>{'user3'}</strong>
    }}
/>
```

hmhealey replied:
Actually, I just discovered that [FormattedList](https://formatjs.io/docs/react-intl/components/#formattedlist) exists. That'd make this much easier
```js
const participants = allParticipants.map((participant) => <strong>{participant}</strong>);

<FormattedMessage
    defaultMessage='Forwarded to {participants}.'
    values={{
        participants: <FormattedList values={participants}/>,
    }}
/>
``` 
Hopefully there's no languages that translate the "Forwarded to" part differently when participants is singular or plural. There's ways around that though

michelengelen replied:
nice find!

matthewbirtch reviewed (approved):
I'm going to approve this to unblock it, because it's working great. I do see that the modal now sits much higher up and all elements are visible/functional now. However, I'm wondering if we should maybe look at centering the modal vertically in a follow-up PR. We can discuss later though.

esethna commented:
@BenCookie95 @hmhealey @Willyfrog kind request to try to complete review soon so we make the feature complete cut for v7.2

rmonga71 commented:
Hey team - was playing around with the test server and noticed that you can't forward a message to someone unless you've previously opened a DM with them. Recorded a [quick vid](https://www.loom.com/share/1a65e42928ae4b4189392afc4ff6042a) using DM with myself as an example (but had the same problem with the test user user-1). Is this expected?

rmonga71 commented:
> The channel doesn't exist if you have never talked with them, so I'd say it is expected. We can try and fix this in a subsequent PR if we want to address it.

Thanks Guillermo for the explanation. 3/5, personally I feel like you should be able to message anyone especially since you have the ability to add a comment for context. Existing DMs could still populate in the recents drop down but as a user I would think that if I wanted to I could forward to anyone with the necessary permission to view the message in question

Willyfrog commented:
> > The channel doesn't exist if you have never talked with them, so I'd say it is expected. We can try and fix this in a subsequent PR if we want to address it.
> 
> Thanks Guillermo for the explanation. 3/5, personally I feel like you should be able to message anyone especially since you have the ability to add a comment for context. Existing DMs could still populate in the recents drop down but as a user I would think that if I wanted to I could forward to anyone with the necessary permission to view the message in question

Sure, can you create a bug ticket for that. What I meant was to allow the PR to be merged and then tackle any edge cases in later PRs, but start using the functionality as soon as possible in community to get feedback and polish based on that.

Willyfrog commented on components/forward_post_modal/forward_post_modal.tsx:
```diff
@@ -0,0 +1,434 @@
+            notificationText = (
+                <FormattedMarkdownMessage
+                    id={'forward_post_modal.notification.private_channel'}
+                    defaultMessage={'This message is from a private channel and can only be shared within **{channel}**'}
```
an easy use case: a message in a thread is sent to the same channel so it is displayed in the center channel and starts a separate thread. Like when a conversation diverges on a thread. This is right now the best we can do (and probably the only thing we'll do) to get some sort of subthreads.

michelengelen commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/217498

rmonga71 commented:
> Sure, can you create a bug ticket for that. What I meant was to allow the PR to be merged and then tackle any edge cases in later PRs, but start using the functionality as soon as possible in community to get feedback and polish based on that.

Agreed with keeping the existing scope of this PR and gathering feedback from Community. Created a ticket for this to be evaluated in the future: https://mattermost.atlassian.net/browse/MM-45688

cwarnermm commented:
Hi @michelengelen! Where can I focus my review efforts in this PR?

matthewbirtch commented:
> Hi @michelengelen! Where can I focus my review efforts in this PR?

@cwarnermm focus on anything related to post forwarding - mainly the forward menu option that appears in the ••• menu on posts and the resulting 'Forward message' modal. There's probably not a lot for you to review though

jgilliam17 commented:
Hi @michelengelen 
seeing few issues
1. White screen when forwarding post from global Threads view
<img width="1169" alt="Screen Shot 2022-07-14 at 9 44 05 AM" src="https://user-images.githubusercontent.com/52937121/178998240-b33f7983-e70b-4a12-83ad-b890d9e27504.png">

2. Adding over the limit text to Add comment field, locks the modal input boxes
<img width="602" alt="Screen Shot 2022-07-13 at 7 21 35 PM" src="https://user-images.githubusercontent.com/52937121/178998381-656b3262-396d-424c-a5b9-8c10af259fde.png">

3.  Sometimes long post forwarding causes few lines of the post to overlap on the forward modal
<img width="601" alt="Screen Shot 2022-07-13 at 7 20 21 PM" src="https://user-images.githubusercontent.com/52937121/178998492-034e9b62-a690-4054-bd03-443ac9ecc22e.png">

4. `new` menu label is not vertically aligned 
<img width="300" alt="Screen Shot 2022-07-13 at 7 09 57 PM" src="https://user-images.githubusercontent.com/52937121/178999532-95a32746-644d-4eab-a811-07b5504a9c67.png">

hmhealey commented on actions/views/posts.js:
```diff
@@ -20,6 +34,66 @@ export function editPost(post) {
+
+        const result = await dispatch(PostActions.createPost(newPost, []));
+
+        // Send to error bar if it's an edit post error about time limit.
```
I think this part can be removed since we're not editing an existing post here

hmhealey commented on components/dot_menu/dot_menu.tsx:
```diff
@@ -421,6 +446,19 @@ export class DotMenuClass extends React.PureComponent<Props, State> {
+
+        const forwardPostItemText = (
+            <span>
+                {Utils.localizeMessage('forward_post_button.label', 'Forward')}
```
These can use `FormattedMessage` since you don't need a raw string

hmhealey commented on components/dot_menu/dot_menu.tsx:
```diff
@@ -421,6 +446,19 @@ export class DotMenuClass extends React.PureComponent<Props, State> {
+            <span>
+                {Utils.localizeMessage('forward_post_button.label', 'Forward')}
+                <Badge variant='success'>
+                    {Utils.localizeMessage('badge.label.new', 'NEW')}
```
Is there any logic for when this should disappear? I'd assume this would go away at some point.

Also, this is a critique of the design, but I would've expected us to use a pulsing green dot for this sort of thing instead of the text NEW
![Screen Shot 2022-07-14 at 10 46 11 AM](https://user-images.githubusercontent.com/3277310/179010402-e191bddb-e327-4df2-813e-9bbb4f15f9a2.png)

michelengelen replied:
this was build as designed... I think the pulsating dot is more for places where there is not much space to actually show a badge.

Right now there is no logic to remove that. I suppose we will just remove the badge in a future dot release.

thoughts @esethna ?

Willyfrog replied:
Usually that pulsating dot gets removed once you click on the feature for the first time

Willyfrog replied:
yeah, I removed the comment because I thought those were your intentions, so I wasn't really adding anything.

hmhealey commented on components/post_view/post_attachment_container/post_attachment_container.tsx:
```diff
@@ -7,33 +7,59 @@ import {useHistory} from 'react-router-dom';
-        }
-    }, [history]);
+        },
+        [history, preventClickAction],
```
I don't think this actually depends on `preventClickAction` any more, so that can be removed

michelengelen commented on components/forward_post_modal/forward_post_modal.tsx:
```diff
@@ -0,0 +1,273 @@
+            handleCancel={onHide}
+            onExited={onHide}
+        >
+            <ForwardPostContext.Provider value={providerValue}>
```
my original idea was to show the text together with the message preview ... like it would be shown when viewed in the channel. I have misread the requirements in that case though, since it was not really wanted and produced more problems than it solved.

I never got to change it back since it works that way perfectly fine ... I am not against changing that to passing down props.

jgilliam17 commented:
One more thing - disregard if this was already mentioned 
There are 2 scroll bars when select channel/people option is longer than modal height making it hard to scroll through the list; you have to scroll the list then move the mouse to the right side and scroll a bit more to see the bottom of the list.
Should we drop the list outside of the modal, similarly to how "add member" modal is designed?
I guess if users use search first, they will see a shorter results list - in that case this may not be an issue.
<img width="616" alt="Screen Shot 2022-07-14 at 7 33 57 PM" src="https://user-images.githubusercontent.com/52937121/179118442-78ac4f22-7073-443e-8bdc-661a2f18ece1.png">

michelengelen commented:
> One more thing - disregard if this was already mentioned There are 2 scroll bars when select channel/people option is longer than modal height making it hard to scroll through the list; you have to scroll the list then move the mouse to the right side and scroll a bit more to see the bottom of the list. Should we drop the list outside of the modal, similarly to how "add member" modal is designed? I guess if users use search first, they will see a shorter results list - in that case this may not be an issue.

Hey @jgilliam17 

Yes, that scrolling problem hasn't been mentioned, but it is something I am already aware of.

There are 2 different approaches to tackle this problem... but simply letting the menu overflow the modal is not easily done (it includes some performance heavy calculations for scrolling I had problems with in the past and I am not sure if this is even possible with the package we are using for the select)... however: If that is the approach that we want to take I can definitely take a look into it.

as for the other option:
measure the height of the modal body and restrict the size of the options-list so that it does not overflow the container and does not add scrollbars to it. 

My latest commit implements that approach, so you can take a look.

in the meantime i will take on the other findings.

---

> Hi @michelengelen seeing few issues
> 
>     1. White screen when forwarding post from global Threads view
> 

Fixed!

> 
>     2. Adding over the limit text to Add comment field, locks the modal input boxes
> 

Fixed!

> 
>     3. Sometimes long post forwarding causes few lines of the post to overlap on the forward modal
> 

this is a separate issue with the message preview component. Ticket is [here](https://mattermost.atlassian.net/browse/MM-45510).

> 
>     4. `new` menu label is not vertically aligned
> 

Fixed!

cwarnermm commented on components/forward_post_modal/forward_post_modal.tsx:
```diff
@@ -0,0 +1,303 @@
+            notificationText = (
+                <FormattedMessage
+                    id='forward_post_modal.notification.dm_or_gm'
+                    defaultMessage='This message is from a private conversation and can only be shared with {participants}'
```
line 3259 has the <strong> tag surrounding {participants} and this instance doesn't.

cwarnermm reviewed (commented):
Thanks, @michelengelen! I've flagged one potential inconsistency inline

hmhealey commented:
To update my review, my only outstanding concern is the NEW badge and it sticking around forever, but we're waiting on @esethna for confirmation on how that should work

esethna commented:
@hmhealey @michelengelen the new badge should not stick around forever. I'm not sure what @matthewbirtch had in mind, but I would imagine we would remove it once the user opens the Forward Message dialog for the first time. We may also want to put a time limitation on it, for example, don't show it for longer than 1 month after the user upgrades (even if they don't click the menu item to forward a message)

Thoughts @michelengelen @hmhealey? 

_______

Also minor thing but the new badge is missing some padding it seems and the corners aren't as rounded as the design:
![image](https://user-images.githubusercontent.com/14333569/179279799-0c9c18a0-3cdd-46fd-9314-a99251dd97c0.png)

![image](https://user-images.githubusercontent.com/14333569/179280014-8c443582-8809-402e-9890-313efc9f2788.png)

esethna commented:
Also the gap between the autocomplete and text box is a bit bigger than designed:

![image](https://user-images.githubusercontent.com/14333569/179281218-79ca6c07-1518-4304-a1bc-beff5d1eb57b.png)


design:
![image](https://user-images.githubusercontent.com/14333569/179281172-dec63a05-3842-46c3-a220-44ab87f41021.png)

michelengelen commented:
>     1. Warning message when sharing a message from private channel, missing channel name formatting
> 

Fixed! 

>     2. Preview component shows hand pointer on hover, and permalinks can be clicked on, navigate to post in the background. I didn't expect this to happen
> 

Fixed by disabling all pointer events on the wrapping component

>     3. Expected? Bots are included on the list - I thought having feedback bot would be convenient as you can share a message that contains/discusses an issue directly with Mattermost. 0/5 on this
> 

I don't know if we want to include bots ... i removed them for now.

>     4. There is a crash when canceling share with feedback bot
>

Fixed by removing bots from the list.

>     5. One more white screen - I was opening Threads, but unsure of the steps before I got there
> 

This is unrelated, but the error is coming from [this selector](https://github.com/mattermost/mattermost-webapp/blob/3975e3c2f61f45c6cef45c54c66c54fae1b444d8/components/threading/global_threads/thread_list/thread_list.tsx#L73) returning `null` and since the return value is immediately destructured it throws an error. This could be easily fixed by giving it an alternate value in case of null 

```
const {total = 0, total_unread_threads: totalUnread} = useSelector(getThreadCountsInCurrentTeam) ?? {};
```

I will open a ticket and PR for this.

>     6. We should remove archived channels from the list
> 

Fixed!

>     7. This may be as designed, just noting here that it wasn't obvious to me as "guest user"
> 

The permalink preview should show that the post is only visible to the user in channel X. Will look into this if it is blocking. @esethna ?

>     8. Forwarding a post from RHS while on Insights is unsuccessful, throws an error
> 

Fixed!

michelengelen commented:
> @hmhealey @michelengelen the new badge should not stick around forever. I'm not sure what @matthewbirtch had in mind, but I would imagine we would remove it once the user opens the Forward Message dialog for the first time. We may also want to put a time limitation on it, for example, don't show it for longer than 1 month after the server is upgraded (even if they don't click the menu item to forward a message)
> 
> Thoughts @michelengelen @hmhealey?

I will add a logic that will remove it once it is clicked (and after a month of latest server upgrade)
 
>     9. Also minor thing but the new badge is missing some padding it seems and the corners aren't as rounded as the design:

Fixed

> 10. Also the gap between the autocomplete and text box is a bit bigger than designed and the autocomplete is being cut off (ie isn't as tall) as designed:
 
Fixed ... I added a logic that restricts the list-height to the maximum height of the modal body ( - input height - margin) since overflowing the container was not possible without breaking the scroll logic for longer forwarded posts and inputs

>     11. Another call out is the scroll bar component seems different than other autocompletes

Fixed

michelengelen commented:
@hmhealey I have added a logic to hide the new badge when it is clicked, but I am struggling to find info on how I can retrieve infos on server version (IMO we should check that as well, since it needs to be a very recent version) and last time updated.

IMO we could just as well remove the badge once the feature is a month old or so.

Thoughts?

hmhealey commented:
I'd be fine just removing it from the code in a future release and just assuming people will have upgraded through one of those versions or will have used the feature by then. As you found, we don't have a good way (at least on the client) of knowing how long a server's been on that specific version. Back when I was working on the NPS survey plugin, that even had to keep track of versions it had seen before to determine that itself.

@esethna Along those lines, I've noticed that this is a bit of a recurring theme where we've cared about how long a server has been on the current version for various things (like these "new feature" indicators or the NPS plugin as I mentioned above). I wonder if it would make sense to get a team to build a proper system for tracking that since it keeps coming up as something we want, but it's always for minor details of features that make it hard to dedicate time to working on. We might also be able to implement features that have a better understanding of if it's a fresh install then (so we can do things like not showing this indicator to users who've never used MM prior to 7.2). Just a thought though.

----

Also, I think https://mattermost.atlassian.net/browse/MM-45781 is a ticket for that error, Michel

hmhealey reviewed (approved):
All that said, my concerns have been addressed now 🙂

esethna commented:
> I have added a logic to hide the new badge when it is clicked, but I am struggling to find info on how I can retrieve infos on server version (IMO we should check that as well, since it needs to be a very recent version) and last time updated.

IMO we could just as well remove the badge once the feature is a month old or so.

@michelengelen @hmhealey The issue with just removing in a future version is that many folks don't upgrade to every release, so it might be missed if someone upgraded, for example, from v6.5 > v7.5

If we're struggling with a way to identify a time based logic (ie remove the badge 1 month after someone first upgrades to a version that supports forwarding a message), then we can just forget that part for now. As long as the new badge will disappear once the menu item is clicked that is probably fine. 

@hmhealey agree it would be nice to have a way to do this though in the future, can you help open a ticket with the technical explanation and share with me so I can add some example use cases?

jgilliam17 commented:
@esethna your thoughts on point 7. https://github.com/mattermost/mattermost-webapp/pull/10613#issuecomment-1185716047

esethna commented:
@jgilliam17 @michelengelen good find re guests users. The permalink preview should not be rendered if the user doesn't have access to the channel where the post exists. We could think about improvements to permalink previews in this case, but I don't think it's blocking.

jgilliam17 commented:
Thanks @michelengelen 
For 3. - Just to clarify - As sysadmin I am not able to search for any DM on the select dropdown. I have user-1 on LHS and have exchanged messages but user-1 is not appearing on the drop down as an option to forward to.

michelengelen commented:
@jgilliam17 the before mentioned issue is now resolved

jgilliam17 commented:
Thanks @michelengelen 
One more thing
The DMs are present on the dropdown, but can't be searched for - return no results

https://user-images.githubusercontent.com/52937121/180013595-badf4065-2be4-4cd6-836e-a895054cf352.mov

michelengelen commented:
> @michelengelen How do I get `new` menu label to show up again?

You should probably create a new user :D

michelengelen commented:
@jgilliam17 there was a mistake on my end ... fixed it! Sorry for the confusion!

jgilliam17 commented:
Unfortunately I am not able to get the [search for DM](https://github.com/mattermost/mattermost-webapp/pull/10613#issuecomment-1190389359) to work on the forward message modal. I'll try updating branch and re-testing

jgilliam17 commented:
/update-branch

jgilliam17 commented:
@michelengelen re-tested - DM search on Forward message modal still not working on the test server

michelengelen commented:
ok, @jgilliam17 I have adressed everything now ... 
1. searching now gets the correct results
2. bots are removed from the list (for now)
3. forwarding to and from every channel works as expected

Could you please re-test and see if you can find anything else ... thanks! ... 🙇🏼

jgilliam17 commented:
Thanks @michelengelen 
Bots are still "present" on the list, albeit invisible 

https://user-images.githubusercontent.com/52937121/180227227-27abf06a-7435-4b79-893a-168e44f8ffc9.mov

jgilliam17 commented:
False alarm - one of the browsers wasn't refreshed 
@michelengelen 
Everything is fixed now - lgtm
I'll create a follow up ticket for guest user experience we discussed previously here.

amyblais commented:
/cherry-pick cloud

mattermod commented:
Cherry pick is scheduled.
